cpp/daal/src/algorithms/dbscan/oneapi/cl_kernels/dbscan_cl_kernels.cl

```diff
@@ -27,161 +27,93 @@
-        point_to_nbr_distance(points, point_id, subgroup_index, power, local_id, subgroup_size, num_features, dist);
-    }
-
-    __kernel void compute_queue_block_distances(__global const algorithmFPType * points, __global const int * queue, int queue_begin, int queue_size,
-                                                int power, int num_features, int num_points, __global algorithmFPType * queue_dist) {
-        const int queue_pos = get_global_id(0);
-        if (queue_pos >= queue_size) return;
-
-        const int point_id                        = queue[queue_begin + queue_pos];
-        const int group_num                       = get_num_sub_groups();
-        const int group_id                        = get_sub_group_id();
-        const int subgroup_size                   = get_sub_group_size();
-        const int local_id                        = get_sub_group_local_id();
-        __global algorithmFPType * cur_block_dist = &queue_dist[queue_pos * num_points];
-        for (int j = group_id; j < num_points; j += group_num)
-        {
-            point_to_nbr_distance(points, point_id, j, power, local_id, subgroup_size, num_features, cur_block_dist);
-        }
-    }
-
-    int get_nbr_status(__global const algorithmFPType * distances, int nbr_id, algorithmFPType eps) { return (distances[nbr_id] <= eps ? 1 : 0); }
-
-    int get_undefined_nbr_status(__global const int * assignments, int point_id, int nbr_id, int is_nbr, int subgroup_offset) {
-        int is_current_point = nbr_id + subgroup_offset == point_id;
-        return (is_nbr > 0 && assignments[nbr_id] == _UNDEFINED_ && !is_current_point ? 1 : 0);
-    }
-
-    __kernel void count_neighbors_by_type(__global const int * assignments, __global const algorithmFPType * distances, int point_id,
-                                          int first_chunk_offset, int chunk_size, int num_points, algorithmFPType eps, __global const int * queue,
-                                          __global int * counters_all_nbrs, __global int * counters_undef_nbrs) {
-        const int subgroup_index        = get_global_id(0) * get_num_sub_groups() + get_sub_group_id();
-        const int subgroup_offset       = subgroup_index * chunk_size;
-        const int subgroup_point_number = num_points - subgroup_offset < chunk_size ? num_points - subgroup_offset : chunk_size;
-        point_id                        = first_chunk_offset < 0 ? point_id : queue[point_id];
-        const int distance_offset       = subgroup_offset + (first_chunk_offset < 0 ? 0 : first_chunk_offset);
-        if (subgroup_index >= num_points) return;
+    __kernel void startNextCluster(int clusterId, int numPoints, int queueEnd, const __global int * cores, __global int * clusters,
+                                   __global int * lastClusterStart, __global int * queue) {
+        // The kernel should be run on a single subgroup
+        if (get_sub_group_id() > 0 || get_global_id(0) > 0) return;
```

Comment:
Is it necessary? We should just spawn workgroups with size of sub groups.

Reply:
I would prefer to have all checks in place.